bazaarvoice/emodb
# Made use of CQL driver configuration persistent

## Github Issue #

[42](https://github.com/bazaarvoice/emodb/issues/42)

## What Are We Doing Here?

There are several things going on in this pull request.  The biggest is the following:

- Changed control of use of CQL or Astyanax driver to be determined by a persistent setting rather than an in-memory valuable which can only be set ephemerally for the duration of the current process.

However, in addition to this there are several other changes:

- Split control of CQL driver for multi-gets and scans into two separately configurable settings, rather than a single setting which controls both.
- Changed setting invalidation to use existing `CacheRegistry`, resulting in more efficient reads for settings by eliminating unnecessary reloads of unmodified settings.
- Separated tasks for setting use of CQL driver versus attributes such as fetch limits for that driver.  Although ideally all of these would use the same settings system the PR was getting long without pulling in unrelated updates, and as such the tasks needed to be separated.
- Removed ability to configure use of CQL or Astyanax driver for single-record reads.  This was necessary because `SettingsManager` uses single-record reads to lookup setting values.  Without this removal a deadlock would occur on the first call to `CqlDataReaderDAO.read()`.  However, the CQL driver has been running for these reads for several weeks without issue, so there is little risk in fully committing to CQL in this case.

Note that with this change the use of CQL or Astyanax is now set across the entire cluster at once; there is no longer any way to AB test CQL on some instances versus Astyanax on others.  However, there has been sufficient testing in anon using both such that the focus at this point is less on comparison testing and more on emergency regression in case of unforeseen issues with the CQL DAO implementation (query errors, memory pressure, who knows?  It's unforeseen).

Finally, the purpose of this PR was to make use of CQL or Astyanax driver persistent in the context of Stash.  As noted in the issue there is no way to call the configuration task on the Stash servers.  However, the Stash servers do read the same setting as persisted by Emo web.  Turning Astyanax on in web prior to the Stash processes starting means that Stash will also use Asytanax.  However, the cache invalidation system used by `CacheRegistry` does not extend to Stash.  If the setting is changed each Stash instance would have to be restarted to force it to re-read the updated setting.

## How to Test and Verify

1. Check out this PR
2. Start EmoDB.
3. Start Stash.  Note the line that says "Setting sor.cassandra.cql.useCqlForScans updated: true"
4. Kill Stash.
5. Use the task to disable CQL scans: `curl -XPOST 'localhost:8081/tasks/sor-cql-driver?use-cql-for-scans=false'`
6. Start Stash again.  Note the line that says "Setting sor.cassandra.cql.useCqlForScans updated: false"

## Risk

Medium.  No one change is terribly large, but on the whole this PR touches a good number of classes.

### Level 

`Medium`

### Required Testing

`Regression`, `Manual`

### Risk Summary

The risks have already been discussed previously, but to summarize the biggest risk is not what's been added but the loss of the ability to control which driver is used on a per-server level since it is now a global system-wide setting.  There are other solutions for this, such as making the setting a blacklist and/or whitelist of IPs, but given that this setting is being used as a fail-safe rather than as a discriminator for AB testing this should be acceptable. 

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
